cmake/CeleritasUtils.cmake

```diff
@@ -43,15 +43,180 @@ macro(celeritas_find_package_config _package)
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_MIDDLE_LIBRARY
+  BRIEF_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+  FULL_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+)
+
+function(celeritas_add_library target)
+  if(NOT BUILD_SHARED_LIBS OR NOT CELERITAS_USE_CUDA)
+    list(SUBLIST ARGV 1 -1 NEWARGV)
+    add_library(${target} ${NEWARGV})
+    add_library(${target}_final ALIAS ${target})
+    add_library(${target}_cuda ALIAS ${target})
+    add_library(${target}_static ALIAS ${target})
+    return()
+  endif()
+
+  list(SUBLIST ARGV 1 -1 NEWARGV)
+
+  add_library(${target}_objects OBJECT ${NEWARGV})
+  add_library(${target}_static STATIC $<TARGET_OBJECTS:${target}_objects>)
+  add_library(${target}_cuda SHARED $<TARGET_OBJECTS:${target}_objects>)
+  # We need to use a dummy file as a library (per cmake) needs to contains
+  # at least one source file.  The real content of the library will be
+  # the cmake_device_link.o resulting from the execution of `nvcc -dlink`
+  # Also non-cuda related test, for example `gtest_detail_Macros`,
+  # will need to be linked again libceleritas_final while a library
+  # that the detends on and that uses Celeritas::Core (for example
+  # libCeleritasTest.so) will need to be linked against `libceleritas_cuda`.
+  # If both the `_cuda` and `_final` contains the `.o` files we would
+  # then have duplicated symbols (Here the symptoms will a crash
+  # during the cuda library initialization rather than a link error).
+  add_library(${target}_final SHARED ../src/base/dummy.cu)
+
+  set_target_properties(${target}_objects PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+    CUDA_SEPARABLE_COMPILATION ON
+    CUDA_RUNTIME_LIBRARY Shared
+  )
+
+  set_target_properties(${target}_cuda PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+    CUDA_SEPARABLE_COMPILATION ON # We really don't want nvlink called.
+    CUDA_RUNTIME_LIBRARY Shared
+    CUDA_RESOLVE_DEVICE_SYMBOLS OFF # We really don't want nvlink called.
+    CELERITAS_CUDA_LIBRARY_TYPE Shared
+    CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
+    CELERITAS_CUDA_STATIC_LIBRARY ${target}_static
+  )
+
+  set_target_properties(${target}_static PROPERTIES
+    LINKER_LANGUAGE CUDA
+    CUDA_SEPARABLE_COMPILATION ON
+    CUDA_RUNTIME_LIBRARY Shared
+    # CUDA_RESOLVE_DEVICE_SYMBOLS OFF # Default for static library
+    CELERITAS_CUDA_LIBRARY_TYPE Static
+    CELERITAS_CUDA_FINAL_LIBRARY ${target}_final
+    CELERITAS_CUDA_MIDDLE_LIBRARY ${target}_cuda
+  )
+
+  set_target_properties(${target}_final PROPERTIES
+    LINKER_LANGUAGE CUDA
+    CUDA_SEPARABLE_COMPILATION ON
+    CUDA_RUNTIME_LIBRARY Shared
+    # CUDA_RESOLVE_DEVICE_SYMBOLS ON # Default for shared library
+    CELERITAS_CUDA_LIBRARY_TYPE Final
+    CELERITAS_CUDA_STATIC_LIBRARY ${target}_static
+    CELERITAS_CUDA_MIDDLE_LIBRARY ${target}_cuda
+  )
```

Comment:
Not necessary for this merge, but the only reason we're using separable compilation in Celeritas is to appease VecGeom; so we should (in the future) only enable RDC if `+cuda+vecgeom`.